### PR TITLE
challenge 2/client-side-application

### PR DESCRIPTION
**Pull request by tomipaul**
Implement client-side application with React and Redux
* Setup webpack
* Implement Actions
* Implement Reducers
* Implement Store
* Implement Components
* Write tests for actions and reducers

### CODE REVIEW

File: client/components/container/Root.jsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import { Provider } from 'react-redux';
+import { BrowserRouter } from 'react-router-dom';
+import configureStore from '../../store/store';
```
Comment (andela-moe) [inline client/components/container/Root.jsx]: A base filename should exactly match the name of its default export. See more details [here](https://github.com/airbnb/javascript#naming--filename-matches-export)

File: client/components/container/Routes.jsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+const Routes = () => {
+  const token = window.localStorage.getItem('auth_token');
+  return (
+    <div>
+      <Route
+        path="/dashboard"
+        render={() => {
+          return (!token) ?
+          (<Redirect to="/" />) :
+          (<LoadDashboard />);
```
Comment (andela-moe) [inline client/components/container/Routes.jsx]: Please use well descriptive names for your components.

**Review by andela-moe — COMMENTED** [review]
Please address comments and other occurences

**Comment by andela-moe** [comment]
Ensure that you resolve your code climate issue and have a good enough GPA for your repo

File: client/components/container/AuthenticateUser.jsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    onSignUpSubmit: ({ username, password, email, phoneNo }) => {
+      dispatch(signUpUser(username, password, email, phoneNo));
+    },
+    onSignInSubmit: ({ username, password }) => {
+      dispatch(signInUser(username, password));
+    }
+  };
+};
+
+const AuthenticateUser = connect(
```
Comment (tomipaul) [inline client/components/container/AuthenticateUser.jsx]:
```
connect(null, {singUpUser, SignInuser})(Authentication)
```

File: client/components/presentation/Authentication/SignUp.jsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+/**
+ * @class SignUp
+ * @extends React.Component
+ */
+class SignUp extends React.Component {
```
Comment (tomipaul) [inline client/components/presentation/Authentication/SignUp.jsx:8]: Convert both signuP and signIn to a functional component

File: client/components/presentation/Common/CommonModal.jsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
+
+/**
+ * @class CommonModal
+ * @extends React.Component
+ */
+class CommonModal extends React.Component {
```
Comment (tomipaul) [inline client/components/presentation/Common/CommonModal.jsx]: Convert to a functional component

File: client/components/presentation/Footer/Priority.jsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const Priority = ({ priorityType, onClick }) => {
+  $(document).ready(() => {
```
Comment (tomipaul) [inline client/components/presentation/Footer/Priority.jsx]: This is bad. change it to use react states to make the update seamless

File: client/components/presentation/Footer/PrioritySelect.jsx [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import Priority from './Priority.jsx';
+
+const PrioritySelect = ({ priorityWasSelected }) => {
+  return (
+    <div className="select-priority-2 col s2">
+      {
+        ['normal', 'urgent', 'critical'].map((priorityType) => {
```
Comment (tomipaul) [inline client/components/presentation/Footer/PrioritySelect.jsx]:
```+        ['normal', 'urgent', 'critical'].map((priorityType) => {
 +          (
 +            <Priority
 +              key={priorityType}
 +              priorityType={priorityType}
 +              onClick={priorityWasSelected}
 +            />
 +          );
 +        })
 +      }
 +    </div>);```

File: client/components/presentation/Main/Index.jsx [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+   * @memberof Main
+   * @returns {void}
+   */
+  hideRightBar() {
+    this.setState({
+      rightBar: false,
+      searchUserView: false,
+      groupMemberView: false
+    });
+    $('.board-body, footer').removeClass('show-right-bar');
```
Comment (tomipaul) [inline client/components/presentation/Main/Index.jsx]: This is bad. change it. className

File: client/components/presentation/Main/SideNav/Index.jsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import UserView from './UserView.jsx';
+import GroupAdd from './GroupAdd.jsx';
+import Group from './Group.jsx';
+
+const SideNav = ({
+  imageLink,
+  username,
+  groups,
```
Comment (tomipaul) [inline client/components/presentation/Main/SideNav/Index.jsx]: groups = {}

File: client/components/presentation/Main/SideNav/Index.jsx [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        username={username}
+      />
+      <GroupAdd />
+      <li>
+        <a className="subheader">GROUPS</a>
+      </li>
+      {
+        (groupIdArray.length !== 0) ?
+        (
+          groupIdArray.map((groupId) => {
```
Comment (tomipaul) [inline client/components/presentation/Main/SideNav/Index.jsx]: `Object.keys(groups)`

File: client/components/presentation/Modals/UploadPictureModal.jsx [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    this.state.footer = (
+      <a
+        href="#!"
+        className="modal-action modal-close waves-effect waves-green btn-flat"
+        onClick={() => {
+          this.props.onSubmit(this.fileinput.files[0]);
+        }}
+      >Upload<i className="material-icons right">send</i></a>
+    );
+    return super.render();
```
Comment (tomipaul) [inline client/components/presentation/Modals/UploadPictureModal.jsx]: This is super bad!.

File: client/reducers/activeGroupUsers.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        action.response.user
+      ];
+    }
+    case REMOVE_USER_FROM_GROUP_SUCCESS: {
+      const groupUsers = [...state];
+      const indexOfRemovedUser = groupUsers
+      .findIndex((user) => {
+        return user.username === action.response.username;
+      });
+      groupUsers.splice(indexOfRemovedUser, 1);
```
Comment (tomipaul) [inline client/reducers/activeGroupUsers.js]: Don't use splice. splice mutate the array. use slice instead.

File: client/reducers/userGroups.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+const userGroups = (state = { groups: {}, groupsById: [] }, action) => {
+  switch (action.type) {
+    case FETCH_USER_GROUPS_SUCCESS: {
+      const groups = action.response.groups
+      .reduce((groupsObject, group) => {
+        groupsObject[group.id] = group;
+        return groupsObject;
+      }, {});
+      const groupsById = action.response.groups
```
Comment (tomipaul) [inline client/reducers/userGroups.js]: ```const groupsById = actions.response.groups.map(group => group.id)```

File: client/components/container/Root.jsx (hunk 1 above)
Comment (tomipaul) [reply client/components/container/Root.jsx]: Okay, I have renamed the file `store.js` to `configureStore.js`

File: client/components/container/AuthenticateUser.jsx (hunk 3 above)
Comment (tomipaul) [reply client/components/container/AuthenticateUser.jsx]: I have changed the mapDispatchToProps argument from a function to an object;
`{ SignInUser, SignUpUser }`

File: client/components/presentation/Footer/PrioritySelect.jsx (hunk 7 above)
Comment (tomipaul) [reply client/components/presentation/Footer/PrioritySelect.jsx]: It gives eslint warnings when I remove return statement

File: client/components/presentation/Main/SideNav/Index.jsx (hunk 9 above)
Comment (tomipaul) [reply client/components/presentation/Main/SideNav/Index.jsx]: Okay, I have effected the change

File: client/reducers/userGroups.js (hunk 13 above)
Comment (tomipaul) [reply client/reducers/userGroups.js]: Okay, I have effected the change

File: client/components/presentation/Footer/Priority.jsx (hunk 6 above)
Comment (tomipaul) [reply client/components/presentation/Footer/Priority.jsx]: The functional component has been converted to a class component and the script moved into componentDidMount of the component.

File: client/components/presentation/Main/SideNav/Index.jsx (hunk 10 above)
Comment (tomipaul) [reply client/components/presentation/Main/SideNav/Index.jsx]: Okay, I have effected the change

File: client/reducers/activeGroupUsers.js (hunk 12 above)
Comment (tomipaul) [reply client/reducers/activeGroupUsers.js]: Code has been refactored. Array.filter method is used to implement logic.

File: client/components/presentation/Authentication/SignUp.jsx (hunk 4 above)
Comment (tomipaul) [reply client/components/presentation/Authentication/SignUp.jsx:8]: Okay, this has been resolved.

File: client/components/presentation/Footer/PrioritySelect.jsx (hunk 7 above)
Comment (tomipaul) [reply client/components/presentation/Footer/PrioritySelect.jsx]: This has been resolved. The right configuration was set in the .eslintrc

File: client/components/presentation/Modals/UploadPictureModal.jsx (hunk 11 above)
Comment (tomipaul) [reply client/components/presentation/Modals/UploadPictureModal.jsx]: Okay, the component has been refactored.

File: client/components/presentation/Common/CommonModal.jsx (hunk 5 above)
Comment (tomipaul) [reply client/components/presentation/Common/CommonModal.jsx]: CommonModal component has been removed. it was aimed at DRYing up the code but in the course of refactoring, I saw it did little to achieve that purpose so I let it go.

File: client/components/container/Routes.jsx (hunk 2 above)
Comment (tomipaul) [reply client/components/container/Routes.jsx]: @andela-moe thanks for the feedback.
```
Footer has been renamed to MessageArea
Home has been renamed to AuthenticationPage
priorityWasSelected was renamed to selectPriority
messageWasChanged was renamed to changeMessage
```

**Pull request merged**